text/0000-expose-scheduler-thread-info-in-stdlib.md

```diff
@@ -0,0 +1,86 @@
+- Feature Name: Expose scheduler thread information in standard library
+- Start Date: 2022-01-12
+- RFC PR: (leave this empty)
+- Pony Issue: (leave this empty)
+
+# Summary
+
+Add a pony primitive that exposes via Pony code information about the scheduler
+and the number of threads available.
+
+# Motivation
+
+Currently, when users want to optimize the parallelizing work based on the maximum number of actors that can be run at a single time, we are pointing them to using the private runtime method `@ponyint_sched_cores()`.
+
+This is problematic for two reasons:
+
+1. We are pointing users at an internal implementation method that we haven't promised not to change.
+2. We are requiring users to muck about with FFI for something that has become somewhat common for users wanting to maximize performance want to know.
+
+By adding a new primitive from which this information can be obtained, we'll be
+setting up a supported API that can be depended on to not change except via the RFC process.
+
+# Detailed design
+
+The new functionality will be added to a new `runtime_info` package that additional information about the runtime can be added to in the future. When this RFC is implemented, the only classes in `runtime_info` will be a single primitive `Scheduler`:
+
+```pony
+use @pony_schedulers[U32]()
+use @pony_active_schedulers[U32]()
+
+primitive Scheduler
+  fun schedulers(auth: RuntimeInfoAuth): U32 =>
+    @pony_schedulers()
+
+  fun active_schedulers(auth: RuntimeInfoAuth): U32 =>
+    @pony_active_schedulers()
+```
+
+The primitive will expose the two existing "what's going on with schedulers" statistics that are available. Currently, only the maximum available number of schedulers is regularly used (as far as I know), but it makes sense to me to expose the current number of active schedulers as well.
+
+We'll replace the existing private functions:
+
+- ponyint_sched_cores
+- ponyint_active_sched_count
+
+with:
+
+- pony_schedulers
+- pony_active_schedulers
+
+which will be marked as `PONY_API`.
+
+Additionally, we will define a single auth required to access runtime information. The auth will be defined in a file called `auth.pony`:
+
+```pony
+type RuntimeInfoAuth is (AmbientAuth | SchedulerInfoAuth)
```

Comment:
This follows a pattern that currently exists. I figured it was good to do that now and be consistent given my expectation that this will go in before any object capability change.

Reply:
It's mainly weird to me that:
- we are treating the name fragments `RuntimeInfo` and `SchedulerInfo` as if they were synonymous in meaning, so there is a redundancy here (they are not as similar in meaning as say `TCPListen` and `TCPListener`).
- we have the name fragment `RuntimeInfo` corresponding to the name of the package (`runtime_info`), but we have the name fragment `SchedulerInfo` corresponding to the primitive `Scheduler` - so where did the `Info` part come from?

Reply:
the package is runtime_info, with a primitive for called "Scheduler" to me that means because this is "runtime info" that "Scheduler" here means it provides "runtime scheduler info", so I called it "SchedulerInfo".

Reply:
The auth could be called "AccessSchedulerInfo".

Reply:
@jemc given that the two of us have agreed to the general shape of an RFC to remove the type unions like this, I'm not real concerned with the naming. I think when we write up the object capability RFC we should change to remove the `RuntimeInfoAuth` and leave only `SchedulerInfoAuth`.

Reply:
What I'd prefer most with the naming is full consistency:
- the library being called `runtime_info`
- the primitive being called `RuntimeInfo`
- the auth primitive being called `RuntimeInfoAuth`

Reply:
I think we should have primitives for different parts of the runtime, not a single RuntimeInfo primitive. I thought that is what we agreed upon in our Zulip conversation.

Reply:
This was discussed on the sync call today. As I mentioned there, I'm still not a fan of this naming within Pony, but it does match the way types are named in many other Pony libraries, including the standard library. And it drives me crazy in all of those other places too, but my complaint is more with the way Pony `use` statements work than the naming conventions itself.